Niloofar / Global ts config - ts configuration for trader and cfd and appstore packages

niloo-fs opened the pull request:
## Changes:

Please include a summary of the change and which issue is fixed below:
-   ...

## When you need to add unit test

-   [ ] If this change disrupt current flow
-   [ ] If this change is adding new flow

## When you need to add integration test

-   [ ] If components from external libraries are being used to define the flow, e.g. @deriv/components
-   [ ] If it relies on a very specific set of props with no default behavior for the current component.

## Test coverage checklist (for reviewer)

-   [ ] Ensure utility / function has a test case 
-   [ ] Ensure all the tests are passing

## Type of change

-   [ ] Bug fix
-   [ ] New feature
-   [ ] Update feature
-   [ ] Refactor code
-   [ ] Translation to code
-   [ ] Translation to crowdin
-   [ ] Script configuration
-   [ ] Improve performance
-   [ ] Style only
-   [ ] Dependency update
-   [ ] Documentation update
-   [ ] Release

niloo-fs commented on packages/appstore/src/types/params.types.ts:
```diff
@@ -1,5 +1,5 @@
 export type TRoute = {
-    component?: React.ReactNode;
+    component?: () => void;
```
Type '() => JSX.Element' is not assignable to type 'ReactNode'.ts(2322)

niloo-fs replied:
I had an issue **Type '() => JSX.Element' is not assignable to type 'ReactNode'.ts(2322)** in **src/constants/routes-config.ts file** (for **TradingHub** component) which failed my build process (after installing @types/react package). one way to fix that was changing the type of the component. the reason I changed the type was that. now I see it would be better to write it like this **() => JSX.Element | React.ReactNode**

niloo-fs replied:
Since react components are functions that return something we need to change the type to a function type which return reactnode (as matin said to support a wider range of types).
Making this small change fixes the issue that we have in our build process.

yashim-deriv commented on packages/appstore/src/types/params.types.ts:
```diff
@@ -1,5 +1,5 @@
 export type TRoute = {
-    component?: React.ReactNode;
+    component?: () => void;
```
This shouldn't be changed. Please find out why `src/constants/routes-config.ts` is throwing an error.

niloo-fs replied:
Since react components are functions that return something we need to change the type to a function type which return reactnode (as matin said to support a wider range of types).
Making this small change fixes the issue that we have in our build process.

niloo-fs replied:
cc: @matin-binary

balakrishna-binary commented on .eslintrc.js:
```diff
@@ -100,11 +100,27 @@ module.exports = {
     },
-  },
-  overrides: [
-    {
-      files: ['**/__tests__/**/*.[jt]s?(x)', '**/?(*.)+(spec|test).[jt]s?(x)'],
-      extends: ['plugin:testing-library/react'],
+    settings: {
+      react: {
+        version: '16',
+      },
+      'import/resolver': {
+        typescript: {}, // this loads <rootdir>/tsconfig.json to eslint
+      },
     },
```
Looks like we already have react prop in settings at line 99. Can we remove the one on line 99 - 102?

niloo-fs replied:
oh you are right. when I updated my branch with upstream master this happened. I'll remove it.

matin-binary reviewed (approved):
LGTM

iman-fs commented:
I noticed that some options in the old ts config are removed/modified in the current version. did you do that on purpose? 

``` "noUnusedLocals" ``` was true but now it's false
``` "noUnusedParameters" ``` was true but now it's false
``` "noImplicitReturns" : true ``` is removed (default value is false)
``` "experimentalDecorators" : true ``` is removed (default value is false)

niloo-fs commented:
> 

I was talking with @hirad-rewok about these, and at the end we decieded to remove unnecessary configs, cause too much configs make the developing process too hard. the reason we removed this was that.

hirad-rewok commented:
> > 
> 
> I was talking with @hirad-rewok about these, and at the end we decieded to remove unnecessary configs, cause too much configs make the developing process too hard. the reason we removed this was that.

We removed this since we have strict set to true which includes a wide range of checks. I double checked this and saw that these four options that are removed are not a part of strict mode and we'd better have them as additional checks. @niloo-fs would you please add them back ?

niloo-fs commented:
> > > 
> > 
> > 
> > I was talking with @hirad-rewok about these, and at the end we decieded to remove unnecessary configs, cause too much configs make the developing process too hard. the reason we removed this was that.
> 
> We removed this since we have strict set to true which includes a wide range of checks. I double checked this and saw that these four options that are removed are not a part of strict mode and we'd better have them as additional checks. @niloo-fs would you please add them back ?

I added these

ali-hosseini-fs reviewed (approved):
LGTM

niloo-fs commented:
> LGTM

Thank u Ako
